src/language/class-modifiers.md

```diff
@@ -0,0 +1,245 @@
+
+class Car extends          // Can be extended
+  Vehicle {
+    int passengers;
+    ...
+}
+
+class MockVehicle          // Can be implemented
+  implements Vehicle {
+    @override
+    void moveForward(int meters) { ... }
+}
+```
+
+If you want your abstract class to appear to be instantiable,
+define a [factory constructor][].
+
+## `base`
+
+Use the `base` modifier to enforce inheritance of a class or mixin's implementation.
+
+The `base` modifier disallows implementation outside of the current library, which
+guarantees:
+
+- The base class constructor is called whenever an instance of a subtype of the
+class is created.
+- That all implemented private members exist in subtypes.
+- Adding new members to a class reduces the risk of breaking subtypes, since all
+subtypes will inherit the new member.
+
+The `base` modifier must be applied transitively to all the subtypes of a class;
+any class which implements or extends a base class must be marked `base` as well. 
+
+```dart
+// Library a.dart
+base class Vehicle {
+  void moveForward(int meters) { ... }
+}
+```
+
+```dart
+// Library b.dart
+import 'a.dart';
+
+var myCar = Vehicle();          // Can be constructed
+
+base class Car extends          // Can be extended
+  Vehicle {
+    int passengers;
+    ...
+}
+
+base class MockVehicle          // ERROR: Cannot be implemented
+  implements Vehicle {
+    @override
+    void moveForward ...
+}
+```
+
+## `interface`
+
+Use the `interface` modifier to define an interface. Interface classes can be
+implemented outside of the current library, but not extended. This guarantees:
+
+- Users of the API are prevented from overriding a subset of methods.
+- A solution to the [fragile base class problem][], or the freedom to change the
+interface's implementation, because there are no inherited member implementations
+outside of the current library whose behavior needs to be preserved. 
+
+```dart
+// Library a.dart
+interface class Vehicle {
+  void moveForward(int meters) { ... }
+}
+```
+
+```dart
+// Library b.dart
+import 'a.dart';
+
+var myCar = Vehicle();     // Can be constructed
+
+class Car extends          // ERROR: Cannot be inherited
+  Vehicle {
+    int passengers;
+    ...
+}
+
+class MockVehicle          // Can be implemented     
+  implements Vehicle {
+    @override
+    void moveForward ...
+}
+```
+
+To define a pure interface that others are free to implement, that cannot be
+inherited and only has abstract members, define an `abstract interface class`.
+
+## `final` 
+
+Use the `final` modifier to prevent subtyping from a class outside of the current
+library, closing the type hierarchy. 
+
+Disallowing both inheritance and implementation means `final` classes provide all
+the same guarantees of [`base`](#base) and [`interface`](#interface) classes.
+
+Final class can be extended or implemented within the
+same library, but like the `base` modifier, any subtypes must be marked `base`.
```

Comment:
Perhaps:
```suggestion
same library, but like the `base` modifier, any subtypes must be marked `base`,
`final`, or `sealed`.
```
The point would be that it is misleading to tell the reader that subtypes of a final class must be `base`. For instance, it makes sense that they would declare every subtype of a given final class `final` as well, thus creating an entirely `final` subtype hierarchy. That hierarchy is completely closed, in the sense that there will never be any subtypes that we don't know about (for instance, if the classes aren't generic then we can use `o.runtimeType == something` to deal with each of those types precisely, which is even stronger than `sealed`). So we shouldn't imply that this is impossible.